### PR TITLE
feat: Help Command Pagination

### PR DESCRIPTION
**Pull request by TAHuntling**
feat: Help Command Pagination

Description: Changed HelpCommand to now be sorted into different pages for commands, which should help to further help users to find the correct commands they are looking for without having to scroll through a list of 92 commands.

Motivation and Context: Recommendation from Aron, and my own experience reading through the /help command list.

How Has This Been Tested?: Running my server and using the /help command will return page 1 of the command list. Running /help 1 will as well, but /help 2 and so forth will return that respective page number of commands (currently 10 commands per page). If a user tries to enter a command higher than the number of pages, it will say it is an invalid page number. Additionally, commands will only show up given the user has the proper GMLevel to use that command.

### CODE REVIEW

**Review by EmosewaMC — COMMENTED** [review]
two comments

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 1]
```diff
@@ -74,39 +73,55 @@ void SlashCommandHandler::HandleChatCommand(const std::u16string& chat, Entity*
...
+		return;
+	}
+
+	// Calculate starting and ending index for commands on the current page
+	size_t startIdx = (page - 1) * pageSize;
+	size_t endIdx = std::min(startIdx + pageSize, accessibleCommands.size());
+
+	// Display commands for the current page
+	feedback << "----- Commands (Page " << page << ") -----";
+	size_t count = 0;
```
[the remark this answers is not in the document]
Comment (TAHuntling) [reply dGame/dUtilities/SlashCommandHandler.cpp:116]: count has been removed

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 2]
```diff
@@ -74,39 +73,55 @@ void SlashCommandHandler::HandleChatCommand(const std::u16string& chat, Entity*
...
+
+	// Check if page number is provided
+	if (!args.empty()) {
+		try {
+			page = std::stoi(args);
+		} catch (const std::exception&) {
+			feedback << "Invalid page number.";
+			GameMessages::SendSlashCommandFeedbackText(entity, GeneralUtils::ASCIIToUTF16(feedback.str()));
+			return;
 		}
```
[the remark this answers is not in the document]
Comment (TAHuntling) [reply dGame/dUtilities/SlashCommandHandler.cpp:101]: Made change in newest push

**Comment by TAHuntling** [comment]
Performed tests after merging with main and the previous sorting changes. Running `/help` will show the list of help terms, alphabetically, and currently is configured to 10 commands at a time to be shown. Running `/help 2` will show page 2 etc. If the user doesnt have a high enough gm, it will alert them that those commands dont exist.

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 3]
```diff
@@ -120,11 +142,6 @@ void SlashCommandHandler::SendAnnouncement(const std::string& title, const std::
 		bitStream.Write<char>(character);
 	}
 
-	bitStream.Write<uint32_t>(message.size());
```
Comment (aronwk-aaron) [inline dGame/dUtilities/SlashCommandHandler.cpp:123]: Why was this removed?

Comment (TAHuntling) [reply dGame/dUtilities/SlashCommandHandler.cpp:123]: it got removed in the merge i did last night, working on it now

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 4]
```diff
@@ -119,11 +141,7 @@ void SlashCommandHandler::SendAnnouncement(const std::string& title, const std::
 	for (auto character : title) {
 		bitStream.Write<char>(character);
 	}
-
 	bitStream.Write<uint32_t>(message.size());
-	for (auto character : message) {
```
Comment (EmosewaMC) [inline dGame/dUtilities/SlashCommandHandler.cpp:124]: Please add back the removed serialization below.

**Review by EmosewaMC — CHANGES_REQUESTED** [review]
- still needs to re add the help<command>
- you removed serialization with will cause issues.

**Comment by TAHuntling** [comment]
Reverted code and reimplemented /help <command>, rather than /<command> help. Alongside keeping pagination

**Review by aronwk-aaron — DISMISSED** [review]

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 5]
```diff
@@ -75,32 +75,90 @@ void SlashCommandHandler::HandleChatCommand(const std::u16string& chat, Entity*
...
+	std::string trimmedArgs = args;
+	trimmedArgs.erase(trimmedArgs.begin(), std::find_if_not(trimmedArgs.begin(), trimmedArgs.end(), [](unsigned char ch) {
+		return std::isspace(ch);
+		}));
+	trimmedArgs.erase(std::find_if_not(trimmedArgs.rbegin(), trimmedArgs.rend(), [](unsigned char ch) {
+		return std::isspace(ch);
+		}).base(), trimmedArgs.end());
+
+	// Check if the argument is a number (for pagination)
+	if (trimmedArgs.empty() || std::all_of(trimmedArgs.begin(), trimmedArgs.end(), ::isdigit)) {
```
Comment (EmosewaMC) [inline dGame/dUtilities/SlashCommandHandler.cpp:90]: Try parse will fail if you try to parse a non number, the logic here can be simplified to that.  Stoi is not used anymore in dlu.

**Review by EmosewaMC — CHANGES_REQUESTED** [review]
Please remove the stoi and use try parse instead.

File: dGame/dUtilities/SlashCommandHandler.cpp [hunk 6]
```diff
@@ -75,32 +73,84 @@ void SlashCommandHandler::HandleChatCommand(const std::u16string& chat, Entity*
...
-			// TODO: Limit displaying commands based on GM level they require
-			if (command.requiredLevel > entity->GetGMLevel()) continue;
-			LOG("Help command: %s", alias.c_str());
+	constexpr size_t pageSize = 10;
+
+	std::string trimmedArgs = args;
+	trimmedArgs.erase(trimmedArgs.begin(), std::find_if_not(trimmedArgs.begin(), trimmedArgs.end(), [](unsigned char ch) {
+		return std::isspace(ch);
+		}));
+	std::optional<int> parsedPage = GeneralUtils::TryParse<int>(trimmedArgs);
```
Comment (EmosewaMC) [inline dGame/dUtilities/SlashCommandHandler.cpp:82]:
```suggestion
	std::optional<uint32_t> parsedPage = GeneralUtils::TryParse<uint32_t>(trimmedArgs);
```

**Review by EmosewaMC — APPROVED** [review]

**Review by aronwk-aaron — APPROVED** [review]
